File: src/renderer/project/report-geometry.ts
// Report screenshot fit geometry, kept pure and electron/DOM-free so the
// no-crop invariant is unit-testable. Same reasoning as export-geometry.ts.
//
// THE BUG THIS EXISTS TO PREVENT (observed: "screenshots are getting cropped
// slightly horizontally in the report view"):
//
// The fit was computed against REPORT_BASE_W (820), but 820 is the width of the
// step CARD, not of the space inside it. `* { box-sizing: border-box }` is global,
// and .rep__bodywrap spends 820 on its own 1px borders and ~1rem side padding, so
// the actual content width is nearer 786. .rep__imgwrap carries max-width:100%, so
// the WRAP was silently clamped to 786 while the IMG kept its inline width of 820,
// and the wrap's overflow:hidden quietly ate the difference.
//
// Horizontal only, which is the clue that named the cause: nothing clamps height,
// so the vertical axis was always fine.
//
// Two rules follow, and the tests assert both:
//   1. fit against the MEASURED content width, never a constant that predates the
//      card's padding.
//   2. the wrap is sized to the image PLUS its own border, so a border-box wrap
//      never has a content box smaller than the image it contains.

import { REPORT_COL_BASE, clampScale } from '../../shared/doc-scale';

/** Max image box at zoom 1, scale 1. Width matches macOS ReportPresentation.baseWidth. */
export const REPORT_BASE_W = REPORT_COL_BASE;
export const REPORT_BASE_H = 600;

/** .rep__imgwrap border width, per side. Must match project.css. */
export const WRAP_BORDER = 1;

export interface ImgDims {
  w: number;
  h: number;
}

export interface ReportFit {
  /** Displayed image size at zoom 1. */
  baseW: number;
  baseH: number;
  /** Outer size for .rep__imgwrap, border included (border-box). */
  wrapW: number;
  wrapH: number;
}

/**
 * Fit a screenshot into the report column.
 *
 * `availW` is the MEASURED content width available to the figure, or null before
 * the first measurement. Null falls back to REPORT_BASE_W, which can over-size by
 * the card's padding for one frame; that is a brief over-size on first paint
 * rather than a persistent crop, and the measured value takes over immediately.
 *
 * Scale is capped at 1: a capture smaller than the column is never blown up.
 */
export function reportFit(
  dims: ImgDims | null,
  availW: number | null,
  zoom = 1,
  scale = 1,
): ReportFit {
  if (!dims || dims.w <= 0 || dims.h <= 0) {
    return { baseW: 0, baseH: 0, wrapW: 0, wrapH: 0 };
  }
  // Subtract the wrap's own border from the budget. Without this the image is
  // sized to the full budget and the border pushes it past the clamp, which is
  // half of how the original crop happened.
  // Both caps scale with the project (#70). Width is normally decided by availW
  // anyway, since the CSS column is already scaled and this measures it; the
  // scaled cap matters when the window is wide enough that the column, not the
  // window, is the limit. HEIGHT has no measured equivalent, so without scaling it
  // a scaled-up capture would grow sideways and then hit an unscaled 600px ceiling.
  const s = clampScale(scale);
  const capW = Math.round(REPORT_BASE_W * s);
  const capH = Math.round(REPORT_BASE_H * s);
  const budgetW = Math.max(1, Math.min(capW, availW ?? capW) - WRAP_BORDER * 2);
  const budgetH = Math.max(1, capH - WRAP_BORDER * 2);

  // Named fitScale, not scale: `scale` is the PROJECT scale parameter. Conflating
  // the two is how a document scale silently becomes an image scale.
  const fitScale = Math.min(budgetW / dims.w, budgetH / dims.h, 1);
  // Floor the image box to whole pixels. A fractional width invites a
  // sub-pixel rounding difference between the wrap's layout and the image's
  // paint, which is a 1px clip of exactly the kind being fixed here.
  const baseW = Math.floor(dims.w * fitScale);
  const baseH = Math.floor(dims.h * fitScale);

  // The box stays fixed at the zoom-1 fit and the image overflows for zoom > 1 so
  // it pans in both axes instead of the box growing taller.
  const boxScale = Math.min(zoom, 1);
  return {
    baseW,
    baseH,
    wrapW: Math.round(baseW * boxScale) + WRAP_BORDER * 2,
    wrapH: Math.round(baseH * boxScale) + WRAP_BORDER * 2,
  };
}

/**
 * The invariant, expressed once so a test can assert it directly: the wrap's
 * CONTENT box (outer minus border, because box-sizing is border-box) must be at
 * least as wide as the image drawn inside it. Any negative slack is a crop.
 */
export function wrapContentSlack(fit: ReportFit): { x: number; y: number } {
  return {
    x: fit.wrapW - WRAP_BORDER * 2 - fit.baseW,
    y: fit.wrapH - WRAP_BORDER * 2 - fit.baseH,
  };
}

File: dotnet/src/ShotAI.Core/Geometry/ReportGeometry.cs
using ShotAI.Core.Json;

namespace ShotAI.Core.Geometry;

/// <summary>An image's natural size in pixels, its orientation applied.</summary>
public readonly record struct ImageSize(double Width, double Height);

/// <summary>
/// <c>ReportFit</c>: a report figure's boxes in whole DIP (INV-REP-8). All zeros is the fit of
/// no image.
/// </summary>
/// <param name="BaseW">The image's displayed width at zoom 1.</param>
/// <param name="BaseH">The image's displayed height at zoom 1.</param>
/// <param name="WrapW">The wrap's outer width, its border included.</param>
/// <param name="WrapH">The wrap's outer height, its border included.</param>
public readonly record struct ReportFit(double BaseW, double BaseH, double WrapW, double WrapH);

/// <summary>
/// The report figure's fit (<c>src/renderer/project/report-geometry.ts</c>, spec 05 2.11 and
/// 7.2), with the same double operations in the same order.
/// </summary>
/// <remarks>
/// The fit is against the width MEASURED for the figure, never a constant that predates the
/// card's padding, and the wrap is the image plus its own border, so its content box is never
/// narrower than the image in it (<c>746b012</c>, INV-REP-4, INV-REP-5).
/// </remarks>
public static class ReportGeometry
{
    /// <summary><c>REPORT_BASE_W</c>: the fit's width cap at scale 1, the export's column (#81).</summary>
    public const int BaseWidth = DocScale.ReportColumnBase;

    /// <summary><c>REPORT_BASE_H</c>: the fit's height cap at scale 1.</summary>
    public const int BaseHeight = 600;

    /// <summary><c>WRAP_BORDER</c>: the wrap's border, per side.</summary>
    public const int WrapBorder = 1;

    /// <summary>
    /// <c>reportFit</c>. Both caps scale with the project (EDGE-REP-4); the image is never
    /// upscaled (INV-REP-6); for a zoom above 1 the wrap keeps its zoom-1 size and the caller
    /// draws the image at <c>BaseW * zoom</c> (INV-REP-7).
    /// </summary>
    /// <param name="image">The natural size; null, or a size that is not positive and finite on both axes, gives all zeros.</param>
    /// <param name="availableWidth">The width measured for the figure, or null before the first measurement (EDGE-REP-2).</param>
    /// <param name="zoom">The figure's zoom.</param>
    /// <param name="scale">The document scale; <see cref="DocScale.Clamp(double)"/> applies.</param>
    /// <remarks>
    /// Electron refuses only a size of zero or less, so a NaN or infinite size gives a NaN fit
    /// there; a decoder never reports one, and here it is no image.
    /// </remarks>
    public static ReportFit Fit(ImageSize? image, double? availableWidth, double zoom = 1, double scale = 1)
    {
        if (image is not { } dims || !IsPositive(dims.Width) || !IsPositive(dims.Height)) return default;
        var s = DocScale.Clamp(scale);
        var capW = JsMath.Round(BaseWidth * s);
        var capH = JsMath.Round(BaseHeight * s);
        var budgetW = Math.Max(1, Math.Min(capW, availableWidth ?? capW) - WrapBorder * 2);
        var budgetH = Math.Max(1, capH - WrapBorder * 2);

        // fitScale, never the project scale: conflating the two is how a document scale
        // silently becomes an image scale.
        var fitScale = Math.Min(Math.Min(budgetW / dims.Width, budgetH / dims.Height), 1);
        var baseW = Math.Floor(dims.Width * fitScale);
        var baseH = Math.Floor(dims.Height * fitScale);
        var boxScale = Math.Min(zoom, 1);
        return new ReportFit(
            baseW,
            baseH,
            JsMath.Round(baseW * boxScale) + WrapBorder * 2,
            JsMath.Round(baseH * boxScale) + WrapBorder * 2);
    }

    /// <summary>
    /// <c>wrapContentSlack</c>: how much wider and taller the wrap's content box is than the
    /// image; a negative value would be a crop (INV-REP-4).
    /// </summary>
    public static (double X, double Y) WrapContentSlack(ReportFit fit) =>
        (fit.WrapW - WrapBorder * 2 - fit.BaseW, fit.WrapH - WrapBorder * 2 - fit.BaseH);

    private static bool IsPositive(double value) => double.IsFinite(value) && value > 0;
}
